frontend/Rhyolite/Frontend/Modal/Base.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Rhyolite.Frontend.Modal.Base where
+
+import Control.Lens (Rewrapped, Wrapped (Unwrapped, _Wrapped'), iso)
+import Control.Monad.Fix (MonadFix)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Primitive (PrimMonad (PrimState, primitive))
+import Control.Monad.Reader (MonadReader)
+import Control.Monad.Ref (MonadAtomicRef, MonadRef)
+import Control.Monad.Trans (MonadTrans (lift))
+import Data.Coerce (coerce)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import Data.Semigroup (First (..))
+import Data.Text (Text)
+import qualified GHCJS.DOM as DOM
+import qualified GHCJS.DOM.EventM as EventM
+import qualified GHCJS.DOM.GlobalEventHandlers as Events
+import Language.Javascript.JSaddle (MonadJSM)
+import Obelisk.Route.Frontend
+import Reflex.Dom.Core
+import Reflex.Host.Class (MonadReflexCreateTrigger)
+
+import Rhyolite.Frontend.Modal.Class (HasModal (ModalM, tellModal))
+
+instance (Reflex t, Monad m) => HasModal t (ModalT t m) where
+  type ModalM (ModalT t m) = m
+  tellModal = ModalT . tellEvent . fmap First
+
+newtype ModalT t m a
+  = ModalT { unModalT :: EventWriterT t (First (Event t () -> m (Event t ()))) m a }
+  deriving
+    ( Functor, Applicative, Monad
+    , MonadFix, MonadIO, MonadRef, MonadAtomicRef, MonadReader r
+    , DomBuilder t, NotReady t, MonadHold t, MonadSample t
+    , PerformEvent t, TriggerEvent t, PostBuild t, HasJS x
+    , MonadReflexCreateTrigger t, MonadQuery t q, Requester t
+    )
+
+instance PrimMonad m => PrimMonad (ModalT t m) where
+  type PrimState (ModalT t m) = PrimState m
+  primitive = lift . primitive
+
+instance Wrapped (ModalT t m a) where
+  type Unwrapped (ModalT t m a) = EventWriterT t (First (Event t () -> m (Event t ()))) m a
+  _Wrapped' = iso coerce coerce
+instance ModalT t m a ~ x => Rewrapped (ModalT t m a) x
+
+instance HasDocument m => HasDocument (ModalT t m)
+instance HasJSContext m => HasJSContext (ModalT t m) where
+  type JSContextPhantom (ModalT t m) = JSContextPhantom m
+  askJSContext = ModalT askJSContext
+#if !defined(ghcjs_HOST_OS)
+instance MonadJSM m => MonadJSM (ModalT t m)
+#endif
+
+instance (Monad m, Routed t r m) => Routed t r (ModalT t m) where
+  askRoute = lift askRoute
```

Comment:
Wish granted meanwhile: https://github.com/obsidiansystems/obelisk/pull/341#issuecomment-455948889

Reply:
I guess we need to bump obelisk to get this?